examples/bridge-app/linux/include/Clusters.h

```diff
@@ -0,0 +1,294 @@
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/AttributeAccessInterface.h>
+#include <app/ConcreteAttributePath.h>
+#include <app/reporting/reporting.h>
+#include <app/util/attribute-storage.h>
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <functional>
+
+static constexpr uint16_t ZCL_DESCRIPTOR_CLUSTER_REVISION              = 1;
+static constexpr uint16_t ZCL_BRIDGED_DEVICE_BASIC_CLUSTER_REVISION    = 1;
+static constexpr uint16_t ZCL_FIXED_LABEL_CLUSTER_REVISION             = 1;
+static constexpr uint16_t ZCL_ON_OFF_CLUSTER_REVISION                  = 4;
+static constexpr uint16_t ZCL_SWITCH_CLUSTER_REVISION                  = 1;
+static constexpr uint16_t ZCL_LEVEL_CONTROL_CLUSTER_REVISION           = 1;
+static constexpr uint16_t ZCL_TEMPERATURE_MEASUREMENT_CLUSTER_REVISION = 1;
+
+// This is the interface to cluster implementations, providing access to manipulate attributes.
+class ClusterImpl
+{
+public:
+    virtual ~ClusterImpl() = default;
+
+    virtual void SetEndpointId(chip::EndpointId id)                          = 0;
+    virtual chip::ClusterId GetClusterId()                                   = 0;
+    virtual chip::Span<const EmberAfAttributeMetadata> GetAllAttributes()    = 0;
+    virtual bool Push(chip::AttributeId attr, chip::TLV::TLVReader & reader) = 0;
+};
+
+// This provides storage for a primitive binary type, eg uintN_t, float, double
+template <typename T, size_t Bytes, EmberAfAttributeType AfType, bool IsFabricScoped = false>
+struct PrimitiveType
+{
+    static constexpr EmberAfAttributeType kMatterType = AfType;
+    // Note that Bytes can be < sizeof(T) due to non-power-of-two byte lengths.
+    static constexpr uint16_t kMaxSize = Bytes;
+    static_assert(Bytes <= sizeof(T), "Incorrect type / byte specification");
+
+    static constexpr bool kIsFabricScoped = IsFabricScoped;
+
+    PrimitiveType() = default;
+    PrimitiveType(T initial) : mValue(initial) {}
+
+    CHIP_ERROR Write(const chip::app::ConcreteDataAttributePath & aPath, chip::app::AttributeValueDecoder & aDecoder)
+    {
+        return aDecoder.Decode(*this);
+    }
+    CHIP_ERROR Read(const chip::app::ConcreteReadAttributePath & aPath, chip::app::AttributeValueEncoder & aEncoder)
+    {
+        return aEncoder.Encode(*this);
+    }
+
+    CHIP_ERROR Decode(chip::TLV::TLVReader & reader) { return reader.Get(mValue); }
+
+    CHIP_ERROR Encode(chip::TLV::TLVWriter & writer, chip::TLV::Tag tag) const { return writer.Put(tag, mValue); }
+
+    void operator=(T v) { mValue = v; }
+    operator T() const { return mValue; }
+
+    T Peek() const { return mValue; }
+
+private:
+    T mValue;
+    static_assert(std::is_standard_layout<T>::value, "PrimitiveType not standard layout!");
+};
+
+// This provides access to a struct. Each struct generates its own access logic.
+template <typename T, bool IsFabricScoped = false>
+struct StructType : public T
```

Comment:
Given my comment in `AddAllAttributes`, I don't think these extra type definitions are warranted, since without the `kMatterType` field, the rest of it is really just a shallow wrapper around the existing templated `AttributeValueDecoder::Decode` and `AttributeValueEncoder::Encode` functions. 

I'd rather we add more types to the set of types that can be passed in there, rather than wrapping them. That would allow you to avoid defining `StructType` and `PrimitiveType` completely.

As for the variable-collection types like octet string and lists, we should create new types backed by `std::string` and `std:::vector` that are usable by `DataModel::Encode` and `DataModel::Decode`. That will be way way more scalable option, and usable by others than just bridge apps.

Reply:
Lists can't be easily handled at the DataModel level because you're missing the context of the ListOperation and/or index

I've got a version where strings are specialized at the DM layer and lists are specialized at the top level and everything else is merged

Reply:
This has now been updated.